rfcs/0127-issues-warnings.md

```diff
@@ -0,0 +1,199 @@
+
+Not all values make sense as issues (i.e. within `meta.issues`): Some warnings may be automatically generated from other `meta` attributes (for example `maintainerless`). New kinds may be added in the future.
+
+## Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+### Package removal
+
+There are two ways issues interact with the removal of packages: Either they get an issue because they are going to be removed, or they are removed because they have an open issue for a prolonged period of time.
+
+- Instead of removing a package directly, it should first get an issue announcing the planned removal. This will allow users to migrate away beforehand. `removal` must be used as `kind` (This will facilitate automation in the future).
+- Before branch-off for a new release, all (leaf) packages with issues that predate the previous branch-off are deemed safe for removal via `throw` (unless stated otherwise). If a package is removed based on its issue, the issue's message becomes part of the new `throw` alias.
+
+### Propagation across transitive dependencies
+
+When a package has an issue, all packages that depend on it will fail to evaluate until that package is ignored or the issue resolved. Sometimes, this is sufficient.
+
+When the issue requires actions on dependents however, it does not sufficiently inform about all packages that need action. Marking all dependents with that issue is not a good idea either though: it would require users to go through some potentially long dependency chains. Instead, only applications, leaf packages or packages with very few dependents should get the issue.
+
+As an example, take `gksu` with the `gksu` → `libgksu` → `libglade` → `python2` dependency chain (for the sake of the example, ignore that it also depends on EOL Gtk 2). Obviously, `python2` should get an issue. As a leaf/application, `gksu` should get one too (it could be the same, or with an adpated message). For the packages in between, it depends on whether they require individual action or not.
+
+### Backporting
+
+New issues generally should not be added to stable branches if possible, and also not be backported to them, since this breaks evaluation. The same rule applies to other changes to a pacakge's `meta` which may generate a warning and thus lead to evaluation failure too. A notable exception are warnings of kind `insecure`, if there is no fix that can be backported.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+- People have voiced strong negative opinions about the prospect of removing packages from nixpkgs at all, especially when they still *technically* work.
+  - We do not want to encourage the use of unmaintained software likely to contain security vulnerabilities, and we do not have the bandwidth to maintain packages deprecated by upstream. Nothing is lost though, because we have complete binary cache coverage of old nixpkgs versions, providing a comparatively easy way to pin old very package versions.
```

Comment:
«*Nothing is lost though*» Have you tried running the bigger ones? Icky impurities and having to rebuild the old world with new versions of externally-talking things is not fun. There is a loss.

«likely to contain security vulnerabilities»: Warnings are flexible. We already have `meta.knownVulnerabilities` too.

Reply:
I have not encountered any major issues so far by pinning down some old dependencies, except for the excessive storage requirements this causes.

The problem about security vulnerabilities, is that they require somebody who cares about you getting the notice. When a package is EOL, this is not guaranteed to happen. Imagine somebody finds a critical bug in Gtk 3. Who will bother to check whether there is a similar exploit in Gtk 2?

Reply:
«Who will bother to check whether there is a similar exploit in Gtk 2?» I'm not sure about Gtk2, but this is a slope; whether a package is abandoned is not binary and we cannot prevent user error. Abandoned software also has less discovered vulnerabilities, which leaves more undiscovered..

«so far by pinning down some old dependencies» – we are improving at that but «excessive storage requirements» are indeed nasty, not to mention build times if you don't «have complete binary cache coverage of old» versions.

-----

The case for making user error harder by doing *something* to tell people vulnerabilities are more likely seems reasonable but dropping seems extreme. Maybe a warning similar to `unmaintained`?